File: packages/react-native-nitro-morph-input/android/src/main/java/com/margelo/nitro/nitromorphinput/HybridMorphInputView.kt
package com.margelo.nitro.nitromorphinput

import android.graphics.Color
import android.os.Handler
import android.os.Looper
import android.view.View
import androidx.annotation.Keep
import com.facebook.proguard.annotations.DoNotStrip
import com.facebook.react.uimanager.ThemedReactContext
import com.margelo.nitro.views.RecyclableView

/**
 * Nitro glue between React props / hybrid methods and [MorphInputView].
 */
@Keep
@DoNotStrip
class HybridMorphInputView(context: ThemedReactContext) : HybridMorphInputViewSpec(), RecyclableView {
  private val inputView = MorphInputView(context)
  private val mainHandler = Handler(Looper.getMainLooper())

  override val view: View
    get() = inputView

  private var isBatching = false
  private var configDirty = true
  private var textDirty = true
  /** The `text` prop value last applied to the field (null = none yet). */
  private var lastAppliedText: String? = null
  /** Native edits so far; JS echoes it back through `mostRecentEventCount`. */
  private var eventCount = 0

  // Read by getText() / getValue() / isFocused(), possibly off the main thread.
  @Volatile private var cachedText = ""
  @Volatile private var cachedValue = Double.NaN
  @Volatile private var cachedFocused = false

  init {
    inputView.onTextChange = { text, value ->
      eventCount += 1
      cachedText = text
      cachedValue = value
      onChangeText?.invoke(text, eventCount.toDouble())
      if (mode == MorphInputMode.NUMBER) onChangeValue?.invoke(value)
    }
    inputView.onFocusChange = { focused ->
      cachedFocused = focused
      onFocusChange?.invoke(focused)
    }
    inputView.onSubmit = { text -> onSubmitEditing?.invoke(text) }
    inputView.onIntrinsicSizeChange = { width, height ->
      onSizeChange?.invoke(width.toDouble(), height.toDouble())
    }
  }

  // region Props

  override var text: String = ""
    set(v) { field = v; markTextDirty() }
  override var mostRecentEventCount: Double = 0.0
    set(v) { field = v; markTextDirty() }
  override var mode: MorphInputMode = MorphInputMode.TEXT
    set(v) { field = v; markConfigDirty() }
  override var fractionDigits: Double = 2.0
    set(v) { field = v; markConfigDirty() }
  override var maxIntegerDigits: Double = 15.0
    set(v) { field = v; markConfigDirty() }
  override var groupingSeparator: String = ","
    set(v) { field = v; markConfigDirty() }
  override var decimalSeparator: String = "."
    set(v) { field = v; markConfigDirty() }
  override var prefix: String = ""
    set(v) { field = v; markConfigDirty() }
  override var suffix: String = ""
    set(v) { field = v; markConfigDirty() }
  override var prefixFontSize: Double = Double.NaN
    set(v) { field = v; markConfigDirty() }
  override var suffixFontSize: Double = Double.NaN
    set(v) { field = v; markConfigDirty() }
  override var affixAlign: MorphInputAffixAlign = MorphInputAffixAlign.BASELINE
    set(v) { field = v; markConfigDirty() }
  override var prefixAlign: MorphInputAffixAlign = MorphInputAffixAlign.BASELINE
    set(v) { field = v; markConfigDirty() }
  override var suffixAlign: MorphInputAffixAlign = MorphInputAffixAlign.BASELINE
    set(v) { field = v; markConfigDirty() }
  override var placeholder: String = ""
    set(v) { field = v; markConfigDirty() }
  override var placeholderColor: Double = Double.NaN
    set(v) { field = v; markConfigDirty() }
  override var duration: Double = 400.0
    set(v) { field = v; markConfigDirty() }
  override var easing: MorphInputEasing = MorphInputEasing.EXPO
    set(v) { field = v; markConfigDirty() }
  override var bounce: Double = 0.15
    set(v) { field = v; markConfigDirty() }
  override var effect: MorphInputEffect = MorphInputEffect.AUTO
    set(v) { field = v; markConfigDirty() }
  override var fontSize: Double = 32.0
    set(v) { field = v; markConfigDirty() }
  override var fontWeight: Double = 400.0
    set(v) { field = v; markConfigDirty() }
  override var fontFamily: String = ""
    set(v) { field = v; markConfigDirty() }
  override var color: Double = Double.NaN
    set(v) { field = v; markConfigDirty() }
  override var textAlign: MorphInputTextAlign = MorphInputTextAlign.LEFT
    set(v) { field = v; markConfigDirty() }
  override var caretColor: Double = Double.NaN
    set(v) { field = v; markConfigDirty() }
  override var selectionColor: Double = Double.NaN
    set(v) { field = v; markConfigDirty() }
  override var caretHidden: Boolean = false
    set(v) { field = v; markConfigDirty() }
  override var adjustsFontSizeToFit: Boolean = false
    set(v) { field = v; markConfigDirty() }
  override var minimumFontScale: Double = 0.5
    set(v) { field = v; markConfigDirty() }
  override var allowFontScaling: Boolean = false
    set(v) { field = v; markConfigDirty() }
  override var maxFontSizeMultiplier: Double = 0.0
    set(v) { field = v; markConfigDirty() }
  override var keyboardType: MorphInputKeyboardType = MorphInputKeyboardType.DEFAULT
    set(v) { field = v; markConfigDirty() }
  override var returnKeyType: MorphInputReturnKeyType = MorphInputReturnKeyType.DEFAULT
    set(v) { field = v; markConfigDirty() }
  override var autoCapitalize: MorphInputAutoCapitalize = MorphInputAutoCapitalize.SENTENCES
    set(v) { field = v; markConfigDirty() }
  override var autoCorrect: Boolean = true
    set(v) { field = v; markConfigDirty() }
  override var editable: Boolean = true
    set(v) { field = v; markConfigDirty() }
  override var autoFocus: Boolean = false
    set(v) { field = v; markConfigDirty() }
  override var maxLength: Double = 0.0
    set(v) { field = v; markConfigDirty() }
  override var transformWorklet: Double = 0.0
    set(v) { field = v; markConfigDirty() }
  override var onChangeTextWorklet: Double = 0.0
    set(v) { field = v; markConfigDirty() }
  override var onChangeValueWorklet: Double = 0.0
    set(v) { field = v; markConfigDirty() }
  override var onChangeText: ((text: String, eventCount: Double) -> Unit)? = null
  override var onChangeValue: ((value: Double) -> Unit)? = null
  override var onFocusChange: ((focused: Boolean) -> Unit)? = null
  override var onSubmitEditing: ((text: String) -> Unit)? = null
  override var onSizeChange: ((width: Double, height: Double) -> Unit)? = null
    set(v) {
      field = v
      onMain { inputView.resendIntrinsicSize() }
    }

  // endregion

  // region Methods

  override fun focus() = onMain { inputView.focus() }

  override fun blur() = onMain { inputView.blur() }

  override fun clear() = onMain {
    flushConfigIfNeeded()
    inputView.clear()
  }

  override fun replaceText(text: String) = onMain {
    flushConfigIfNeeded()
    inputView.replaceText(text)
  }

  override fun setValue(value: Double) = onMain {
    flushConfigIfNeeded()
    inputView.setValue(value)
  }

  override fun currentText(): String = cachedText

  override fun getValue(): Double = cachedValue

  override fun isFocused(): Boolean = cachedFocused

  // endregion

  // region Lifecycle

  override fun beforeUpdate() {
    isBatching = true
  }

  override fun afterUpdate() {
    isBatching = false
    commit()
  }

  override fun onDropView() {
    onMain { inputView.stopAnimation() }
  }

  /**
   * Fabric is about to reuse this view for another element: forget every prop,
   * the text and the edit count. Nitro re-applies the new element's props next.
   */
  override fun prepareForRecycle() {
    isBatching = true
    text = ""
    mostRecentEventCount = 0.0
    mode = MorphInputMode.TEXT
    fractionDigits = 2.0
    maxIntegerDigits = 15.0
    groupingSeparator = ","
    decimalSeparator = "."
    prefix = ""
    suffix = ""
    prefixFontSize = Double.NaN
    suffixFontSize = Double.NaN
    affixAlign = MorphInputAffixAlign.BASELINE
    prefixAlign = MorphInputAffixAlign.BASELINE
    suffixAlign = MorphInputAffixAlign.BASELINE
    placeholder = ""
    placeholderColor = Double.NaN
    duration = 400.0
    easing = MorphInputEasing.EXPO
    bounce = 0.15
    effect = MorphInputEffect.AUTO
    fontSize = 32.0
    fontWeight = 400.0
    fontFamily = ""
    color = Double.NaN
    textAlign = MorphInputTextAlign.LEFT
    caretColor = Double.NaN
    selectionColor = Double.NaN
    caretHidden = false
    adjustsFontSizeToFit = false
    minimumFontScale = 0.5
    allowFontScaling = false
    maxFontSizeMultiplier = 0.0
    keyboardType = MorphInputKeyboardType.DEFAULT
    returnKeyType = MorphInputReturnKeyType.DEFAULT
    autoCapitalize = MorphInputAutoCapitalize.SENTENCES
    autoCorrect = true
    editable = true
    autoFocus = false
    maxLength = 0.0
    transformWorklet = 0.0
    onChangeTextWorklet = 0.0
    onChangeValueWorklet = 0.0
    onChangeText = null
    onChangeValue = null
    onFocusChange = null
    onSubmitEditing = null
    onSizeChange = null
    lastAppliedText = null
    eventCount = 0
    cachedText = ""
    cachedValue = Double.NaN
    cachedFocused = false
    configDirty = true
    textDirty = true
    isBatching = false
    onMain { inputView.resetForRecycle() }
  }

  // endregion

  // region Batching

  private fun markConfigDirty() {
    configDirty = true
    commitIfNeeded()
  }

  private fun markTextDirty() {
    textDirty = true
    commitIfNeeded()
  }

  private fun commitIfNeeded() {
    if (!isBatching) commit()
  }

  private fun commit() {
    onMain {
      inputView.batch {
        flushConfigIfNeeded()
        flushTextIfNeeded()
      }
    }
  }

  /**
   * The `text` / `mostRecentEventCount` handshake: a new `text` is applied only
   * once JS has seen every native edit (a stale count means the user typed
   * since, and the field already shows what they typed).
   */
  private fun flushTextIfNeeded() {
    if (!textDirty) return
    textDirty = false
    val value = text
    if (lastAppliedText == value) return
    if (mostRecentEventCount < eventCount) return
    lastAppliedText = value
    inputView.applyText(value)
    cachedText = inputView.text
    cachedValue = inputView.currentValue()
  }

  private fun flushConfigIfNeeded() {
    if (!configDirty) return
    configDirty = false

    val size = finite(fontSize, 32.0).toFloat()
    inputView.format = MorphInputView.Format(
      mode = if (mode == MorphInputMode.NUMBER) MorphInputView.Mode.NUMBER else MorphInputView.Mode.TEXT,
      fractionDigits = clampInt(fractionDigits, 0, 9, 2),
      maxIntegerDigits = clampInt(maxIntegerDigits, 1, 30, 15),
      groupingSeparator = groupingSeparator,
      decimalSeparator = decimalSeparator,
      prefix = prefix,
      suffix = suffix,
      placeholder = placeholder,
    )
    inputView.typography = MorphInputView.Typography(
      fontSize = size,
      prefixFontSize = finite(prefixFontSize, size.toDouble()).toFloat(),
      suffixFontSize = finite(suffixFontSize, size.toDouble()).toFloat(),
      fontWeight = clampInt(fontWeight, 100, 900, 400),
      fontFamily = fontFamily.ifEmpty { null },
      color = colorFromARGB(color),
      placeholderColor = colorFromARGB(placeholderColor),
      prefixAlign = mapAffixAlign(prefixAlign),
      suffixAlign = mapAffixAlign(suffixAlign),
      adjustsFontSizeToFit = adjustsFontSizeToFit,
      minimumFontScale = finite(minimumFontScale, 0.5).coerceIn(0.05, 1.0).toFloat(),
      allowFontScaling = allowFontScaling,
      maxFontSizeMultiplier = Math.max(0.0, finite(maxFontSizeMultiplier, 0.0)).toFloat(),
    )
    inputView.worklets = MorphInputView.Worklets(
      transform = clampInt(transformWorklet, 0, Int.MAX_VALUE, 0),
      onChangeText = clampInt(onChangeTextWorklet, 0, Int.MAX_VALUE, 0),
      onChangeValue = clampInt(onChangeValueWorklet, 0, Int.MAX_VALUE, 0),
    )
    inputView.timing = MorphInputView.Timing(
      durationMs = Math.max(0.0, finite(duration, 400.0)).toLong(),
      easing = when (easing) {
        MorphInputEasing.EASEOUT -> MorphInputView.Easing.EASE_OUT
        MorphInputEasing.EASEINOUT -> MorphInputView.Easing.EASE_IN_OUT
        MorphInputEasing.LINEAR -> MorphInputView.Easing.LINEAR
        MorphInputEasing.SPRING -> MorphInputView.Easing.SPRING
        MorphInputEasing.EXPO -> MorphInputView.Easing.EXPO
      },
      bounce = finite(bounce, 0.15).coerceIn(0.0, 1.0),
      effect = when (effect) {
        MorphInputEffect.SLIDE -> MorphInputView.Effect.SLIDE
        MorphInputEffect.FADE -> MorphInputView.Effect.FADE
        MorphInputEffect.AUTO -> MorphInputView.Effect.AUTO
      },
    )
    inputView.keyboard = MorphInputView.Keyboard(
      keyboardType = when (keyboardType) {
        MorphInputKeyboardType.DEFAULT -> MorphInputView.KeyboardType.DEFAULT
        MorphInputKeyboardType.NUMBER_PAD -> MorphInputView.KeyboardType.NUMBER_PAD
        MorphInputKeyboardType.DECIMAL_PAD -> MorphInputView.KeyboardType.DECIMAL_PAD
        MorphInputKeyboardType.NUMERIC -> MorphInputView.KeyboardType.NUMERIC
        MorphInputKeyboardType.EMAIL_ADDRESS -> MorphInputView.KeyboardType.EMAIL_ADDRESS
        MorphInputKeyboardType.PHONE_PAD -> MorphInputView.KeyboardType.PHONE_PAD
        MorphInputKeyboardType.URL -> MorphInputView.KeyboardType.URL
        MorphInputKeyboardType.ASCII_CAPABLE -> MorphInputView.KeyboardType.ASCII_CAPABLE
        MorphInputKeyboardType.NUMBERS_AND_PUNCTUATION -> MorphInputView.KeyboardType.NUMBERS_AND_PUNCTUATION
      },
      returnKeyType = when (returnKeyType) {
        MorphInputReturnKeyType.DEFAULT -> MorphInputView.ReturnKeyType.DEFAULT
        MorphInputReturnKeyType.DONE -> MorphInputView.ReturnKeyType.DONE
        MorphInputReturnKeyType.GO -> MorphInputView.ReturnKeyType.GO
        MorphInputReturnKeyType.NEXT -> MorphInputView.ReturnKeyType.NEXT
        MorphInputReturnKeyType.SEARCH -> MorphInputView.ReturnKeyType.SEARCH
        MorphInputReturnKeyType.SEND -> MorphInputView.ReturnKeyType.SEND
      },
      autoCapitalize = when (autoCapitalize) {
        MorphInputAutoCapitalize.NONE -> MorphInputView.AutoCapitalize.NONE
        MorphInputAutoCapitalize.SENTENCES -> MorphInputView.AutoCapitalize.SENTENCES
        MorphInputAutoCapitalize.WORDS -> MorphInputView.AutoCapitalize.WORDS
        MorphInputAutoCapitalize.CHARACTERS -> MorphInputView.AutoCapitalize.CHARACTERS
      },
      autoCorrect = autoCorrect,
      editable = editable,
      autoFocus = autoFocus,
      maxLength = clampInt(maxLength, 0, Int.MAX_VALUE, 0),
    )
    inputView.caret = MorphInputView.Caret(
      color = colorFromARGB(caretColor),
      selectionColor = colorFromARGB(selectionColor),
      hidden = caretHidden,
    )
    inputView.alignment = when (textAlign) {
      MorphInputTextAlign.CENTER -> MorphInputView.Alignment.CENTER
      MorphInputTextAlign.RIGHT -> MorphInputView.Alignment.RIGHT
      MorphInputTextAlign.LEFT -> MorphInputView.Alignment.LEFT
    }
    // A format change may have re-normalised the field's text.
    cachedText = inputView.text
    cachedValue = inputView.currentValue()
  }

  // endregion

  private fun mapAffixAlign(align: MorphInputAffixAlign): MorphInputView.AffixAlign = when (align) {
    MorphInputAffixAlign.CENTER -> MorphInputView.AffixAlign.CENTER
    MorphInputAffixAlign.TOP -> MorphInputView.AffixAlign.TOP
    MorphInputAffixAlign.BOTTOM -> MorphInputView.AffixAlign.BOTTOM
    MorphInputAffixAlign.BASELINE -> MorphInputView.AffixAlign.BASELINE
  }

  private fun onMain(block: () -> Unit) {
    if (Looper.myLooper() == Looper.getMainLooper()) block() else mainHandler.post(block)
  }

  private fun finite(value: Double, fallback: Double): Double = if (value.isFinite()) value else fallback

  private fun clampInt(value: Double, lower: Int, upper: Int, fallback: Int): Int {
    if (!value.isFinite()) return fallback
    return Math.rint(value).coerceIn(lower.toDouble(), upper.toDouble()).toInt()
  }

  private fun colorFromARGB(value: Double): Int? {
    if (!value.isFinite()) return null
    val bits = value.toLong().toInt()
    return Color.argb((bits ushr 24) and 0xff, (bits ushr 16) and 0xff, (bits ushr 8) and 0xff, bits and 0xff)
  }
}

File: packages/react-native-nitro-morph-input/ios/HybridMorphInputView.swift
//
//  HybridMorphInputView.swift
//  NitroMorphInput
//
//  Nitro glue between React props / hybrid methods and `MorphInputView`.
//

import Foundation
import NitroModules
import UIKit

final class HybridMorphInputView: HybridMorphInputViewSpec, RecyclableView {
  private let inputView = MorphInputView()

  var view: UIView { inputView }

  private var isBatching = false
  private var configDirty = true
  /// The `text` prop as last applied; nil until the first application.
  private var lastAppliedText: String?
  private var textDirty = true

  /// What `getText()` / `getValue()` / `isFocused()` answer with. Nitro may
  /// call them off the main thread, so the view's state is mirrored here.
  private struct Snapshot {
    var text = ""
    var value = Double.nan
    var focused = false
    var eventCount = 0
  }
  private let snapshotLock = NSLock()
  private var snapshot = Snapshot()

  override init() {
    super.init()
    inputView.onTextChange = { [weak self] text, value, eventCount, reason in
      guard let self else { return }
      self.snapshotLock.lock()
      self.snapshot.text = text
      self.snapshot.value = value
      self.snapshot.eventCount = eventCount
      self.snapshotLock.unlock()
      guard reason != .prop else { return }
      self.onChangeText?(text, Double(eventCount))
      if self.mode == .number {
        self.onChangeValue?(value)
      }
    }
    inputView.onFocusChange = { [weak self] focused in
      guard let self else { return }
      self.snapshotLock.lock()
      self.snapshot.focused = focused
      self.snapshotLock.unlock()
      self.onFocusChange?(focused)
    }
    inputView.onSubmit = { [weak self] text in
      self?.onSubmitEditing?(text)
    }
    inputView.onIntrinsicSizeChange = { [weak self] size in
      self?.onSizeChange?(Double(size.width), Double(size.height))
    }
  }

  // MARK: - Props

  var text: String = "" { didSet { markTextDirty() } }
  var mostRecentEventCount: Double = 0 { didSet { markTextDirty() } }
  var mode: MorphInputMode = .text { didSet { markConfigDirty() } }
  var fractionDigits: Double = 2 { didSet { markConfigDirty() } }
  var maxIntegerDigits: Double = 15 { didSet { markConfigDirty() } }
  var groupingSeparator: String = "," { didSet { markConfigDirty() } }
  var decimalSeparator: String = "." { didSet { markConfigDirty() } }
  var prefix: String = "" { didSet { markConfigDirty() } }
  var suffix: String = "" { didSet { markConfigDirty() } }
  var prefixFontSize: Double = 32 { didSet { markConfigDirty() } }
  var suffixFontSize: Double = 32 { didSet { markConfigDirty() } }
  var affixAlign: MorphInputAffixAlign = .baseline { didSet { markConfigDirty() } }
  var prefixAlign: MorphInputAffixAlign = .baseline { didSet { markConfigDirty() } }
  var suffixAlign: MorphInputAffixAlign = .baseline { didSet { markConfigDirty() } }
  var placeholder: String = "" { didSet { markConfigDirty() } }
  var placeholderColor: Double = .nan { didSet { markConfigDirty() } }
  var duration: Double = 400 { didSet { markConfigDirty() } }
  var easing: MorphInputEasing = .expo { didSet { markConfigDirty() } }
  var bounce: Double = 0.15 { didSet { markConfigDirty() } }
  var effect: MorphInputEffect = .auto { didSet { markConfigDirty() } }
  var fontSize: Double = 32 { didSet { markConfigDirty() } }
  var fontWeight: Double = 400 { didSet { markConfigDirty() } }
  var fontFamily: String = "" { didSet { markConfigDirty() } }
  var color: Double = .nan { didSet { markConfigDirty() } }
  var textAlign: MorphInputTextAlign = .left { didSet { markConfigDirty() } }
  var caretColor: Double = .nan { didSet { markConfigDirty() } }
  var selectionColor: Double = .nan { didSet { markConfigDirty() } }
  var caretHidden: Bool = false { didSet { markConfigDirty() } }
  var adjustsFontSizeToFit: Bool = false { didSet { markConfigDirty() } }
  var minimumFontScale: Double = 0.5 { didSet { markConfigDirty() } }
  var allowFontScaling: Bool = false { didSet { markConfigDirty() } }
  var maxFontSizeMultiplier: Double = 0 { didSet { markConfigDirty() } }
  var keyboardType: MorphInputKeyboardType = .default { didSet { markConfigDirty() } }
  var returnKeyType: MorphInputReturnKeyType = .default { didSet { markConfigDirty() } }
  var autoCapitalize: MorphInputAutoCapitalize = .sentences { didSet { markConfigDirty() } }
  var autoCorrect: Bool = true { didSet { markConfigDirty() } }
  var editable: Bool = true { didSet { markConfigDirty() } }
  var autoFocus: Bool = false { didSet { markConfigDirty() } }
  var maxLength: Double = 0 { didSet { markConfigDirty() } }
  var onChangeText: ((String, Double) -> Void)?
  var onChangeValue: ((Double) -> Void)?
  var onFocusChange: ((Bool) -> Void)?
  var onSubmitEditing: ((String) -> Void)?
  var onSizeChange: ((Double, Double) -> Void)? {
    didSet { onMain { self.inputView.resendIntrinsicSize() } }
  }

  // MARK: - Methods

  func focus() throws {
    onMain {
      self.flushConfigIfNeeded()
      self.inputView.focus()
    }
  }

  func blur() throws {
    onMain { self.inputView.blur() }
  }

  func clear() throws {
    onMain {
      self.flushConfigIfNeeded()
      self.inputView.clear(reason: .method)
    }
  }

  func replaceText(text: String) throws {
    onMain {
      self.flushConfigIfNeeded()
      self.inputView.setText(text, reason: .method)
    }
  }

  func setValue(value: Double) throws {
    onMain {
      self.flushConfigIfNeeded()
      self.inputView.setValue(value, reason: .method)
    }
  }

  func currentText() throws -> String {
    snapshotLock.lock()
    defer { snapshotLock.unlock() }
    return snapshot.text
  }

  func getValue() throws -> Double {
    snapshotLock.lock()
    defer { snapshotLock.unlock() }
    return snapshot.value
  }

  func isFocused() throws -> Bool {
    snapshotLock.lock()
    defer { snapshotLock.unlock() }
    return snapshot.focused
  }

  // MARK: - Lifecycle

  func beforeUpdate() {
    isBatching = true
  }

  func afterUpdate() {
    isBatching = false
    commit()
  }

  func onDropView() {
    onMain { self.inputView.stopAnimation() }
  }

  /// Fabric is about to reuse this view for another element: forget every prop
  /// and all state. Nitro re-applies the new element's props next.
  func prepareForRecycle() {
    isBatching = true
    text = ""
    mostRecentEventCount = 0
    mode = .text
    fractionDigits = 2
    maxIntegerDigits = 15
    groupingSeparator = ","
    decimalSeparator = "."
    prefix = ""
    suffix = ""
    prefixFontSize = 32
    suffixFontSize = 32
    affixAlign = .baseline
    prefixAlign = .baseline
    suffixAlign = .baseline
    placeholder = ""
    placeholderColor = .nan
    duration = 400
    easing = .expo
    bounce = 0.15
    effect = .auto
    fontSize = 32
    fontWeight = 400
    fontFamily = ""
    color = .nan
    textAlign = .left
    caretColor = .nan
    selectionColor = .nan
    caretHidden = false
    adjustsFontSizeToFit = false
    minimumFontScale = 0.5
    allowFontScaling = false
    maxFontSizeMultiplier = 0
    keyboardType = .default
    returnKeyType = .default
    autoCapitalize = .sentences
    autoCorrect = true
    editable = true
    autoFocus = false
    maxLength = 0
    onChangeText = nil
    onChangeValue = nil
    onFocusChange = nil
    onSubmitEditing = nil
    onSizeChange = nil
    lastAppliedText = nil
    configDirty = true
    textDirty = true
    isBatching = false
    snapshotLock.lock()
    snapshot = Snapshot()
    snapshotLock.unlock()
    onMain { self.inputView.resetForRecycle() }
  }

  // MARK: - Batching

  private func markConfigDirty() {
    configDirty = true
    commitIfNeeded()
  }

  private func markTextDirty() {
    textDirty = true
    commitIfNeeded()
  }

  private func commitIfNeeded() {
    if !isBatching { commit() }
  }

  private func commit() {
    onMain {
      self.flushConfigIfNeeded()
      self.applyTextIfNeeded()
    }
  }

  /// The `text` prop handshake: apply it when it differs from the last applied
  /// value and JS has seen every native edit (`mostRecentEventCount` caught up
  /// with the native count), like React Native's own `TextInput`.
  private func applyTextIfNeeded() {
    guard textDirty else { return }
    textDirty = false
    guard text != lastAppliedText else { return }
    snapshotLock.lock()
    let nativeCount = snapshot.eventCount
    snapshotLock.unlock()
    guard mostRecentEventCount.isFinite, Int(mostRecentEventCount) >= nativeCount else { return }
    lastAppliedText = text
    inputView.setText(text, reason: .prop)
  }

  private func flushConfigIfNeeded() {
    guard configDirty else { return }
    configDirty = false

    var format = MorphInputView.Format()
    format.mode = mode == .number ? .number : .text
    format.fractionDigits = Self.clampInt(fractionDigits, 0, 9, fallback: 2)
    format.maxIntegerDigits = Self.clampInt(maxIntegerDigits, 1, 15, fallback: 15)
    format.groupingSeparator = groupingSeparator
    format.decimalSeparator = decimalSeparator.isEmpty ? "." : decimalSeparator
    format.prefix = prefix
    format.suffix = suffix
    format.placeholder = placeholder
    format.maxLength = Self.clampInt(maxLength, 0, 1_000_000, fallback: 0)

    var typography = MorphInputView.Typography()
    typography.fontSize = CGFloat(fontSize.isFinite && fontSize > 0 ? fontSize : 32)
    typography.prefixFontSize = prefixFontSize.isFinite && prefixFontSize > 0 ? CGFloat(prefixFontSize) : nil
    typography.suffixFontSize = suffixFontSize.isFinite && suffixFontSize > 0 ? CGFloat(suffixFontSize) : nil
    typography.fontWeight = CGFloat(fontWeight.isFinite ? fontWeight : 400)
    typography.fontFamily = fontFamily.isEmpty ? nil : fontFamily
    typography.color = Self.color(fromARGB: color) ?? .label
    typography.placeholderColor = Self.color(fromARGB: placeholderColor) ?? .placeholderText
    typography.prefixAlign = Self.mapAffixAlign(prefixAlign)
    typography.suffixAlign = Self.mapAffixAlign(suffixAlign)
    typography.adjustsFontSizeToFit = adjustsFontSizeToFit
    typography.minimumFontScale = CGFloat(min(1, max(0.05, minimumFontScale.isFinite ? minimumFontScale : 0.5)))
    typography.allowFontScaling = allowFontScaling
    typography.maxFontSizeMultiplier = CGFloat(max(0, maxFontSizeMultiplier.isFinite ? maxFontSizeMultiplier : 0))

    var timing = MorphInputView.Timing()
    timing.duration = max(0, (duration.isFinite ? duration : 400) / 1000)
    timing.easing = Self.mapEasing(easing)
    timing.bounce = min(1, max(0, bounce.isFinite ? bounce : 0.15))
    timing.effect = Self.mapEffect(effect)

    var traits = MorphInputView.Traits()
    traits.keyboardType = Self.mapKeyboardType(keyboardType)
    traits.returnKeyType = Self.mapReturnKeyType(returnKeyType)
    traits.autocapitalization = Self.mapAutoCapitalize(autoCapitalize)
    traits.autocorrect = autoCorrect
    traits.editable = editable
    traits.autoFocus = autoFocus
    traits.caretHidden = caretHidden
    traits.caretColor = Self.color(fromARGB: caretColor)
    traits.selectionColor = Self.color(fromARGB: selectionColor)

    inputView.typography = typography
    inputView.format = format
    inputView.timing = timing
    inputView.traits = traits
    inputView.alignment = Self.mapAlignment(textAlign)
  }

  // MARK: - Helpers

  private func onMain(_ block: @escaping () -> Void) {
    if Thread.isMainThread {
      block()
    } else {
      DispatchQueue.main.async(execute: block)
    }
  }

  private static func clampInt(_ value: Double, _ lower: Int, _ upper: Int, fallback: Int) -> Int {
    guard value.isFinite else { return fallback }
    return min(upper, max(lower, Int(value.rounded())))
  }

  /// A processed React Native color (ARGB packed in a double); NaN = platform default.
  private static func color(fromARGB value: Double) -> UIColor? {
    guard value.isFinite else { return nil }
    let bits = UInt32(truncatingIfNeeded: Int64(value))
    let a = CGFloat((bits >> 24) & 0xff) / 255
    let r = CGFloat((bits >> 16) & 0xff) / 255
    let g = CGFloat((bits >> 8) & 0xff) / 255
    let b = CGFloat(bits & 0xff) / 255
    return UIColor(red: r, green: g, blue: b, alpha: a)
  }

  private static func mapEasing(_ easing: MorphInputEasing) -> Int32 {
    switch easing {
    case .expo: return 0
    case .easeout: return 1
    case .easeinout: return 2
    case .linear: return 3
    case .spring: return 4
    default: return 0
    }
  }

  private static func mapEffect(_ effect: MorphInputEffect) -> Int32 {
    switch effect {
    case .auto: return 0
    case .slide: return 1
    case .fade: return 2
    default: return 0
    }
  }

  private static func mapAffixAlign(_ align: MorphInputAffixAlign) -> MorphInputView.AffixAlign {
    switch align {
    case .baseline: return .baseline
    case .center: return .center
    case .top: return .top
    case .bottom: return .bottom
    default: return .baseline
    }
  }

  private static func mapAlignment(_ align: MorphInputTextAlign) -> MorphInputView.Alignment {
    switch align {
    case .left: return .left
    case .center: return .center
    case .right: return .right
    default: return .left
    }
  }

  private static func mapKeyboardType(_ type: MorphInputKeyboardType) -> UIKeyboardType {
    switch type {
    case .default: return .default
    case .numberPad: return .numberPad
    case .decimalPad: return .decimalPad
    case .numeric: return .numbersAndPunctuation
    case .emailAddress: return .emailAddress
    case .phonePad: return .phonePad
    case .url: return .URL
    case .asciiCapable: return .asciiCapable
    case .numbersAndPunctuation: return .numbersAndPunctuation
    default: return .default
    }
  }

  private static func mapReturnKeyType(_ type: MorphInputReturnKeyType) -> UIReturnKeyType {
    switch type {
    case .default: return .default
    case .done: return .done
    case .go: return .go
    case .next: return .next
    case .search: return .search
    case .send: return .send
    default: return .default
    }
  }

  private static func mapAutoCapitalize(_ type: MorphInputAutoCapitalize) -> UITextAutocapitalizationType {
    switch type {
    case .none: return .none
    case .sentences: return .sentences
    case .words: return .words
    case .characters: return .allCharacters
    default: return .sentences
    }
  }
}
